File: 33. Search in Rotated Sorted Array/33. Search in Rotated Sorted Array.ts
function search(nums: number[], target: number): number {
    let left: number = 0;
    let right: number = nums.length - 1;
    
    while (left <= right) {
        const mid: number = left + Math.floor((right - left) / 2);
        
        if (nums[mid] === target) {
            return mid;
        }
        
        // Check if the left half is sorted
        if (nums[left] <= nums[mid]) {
            if (target >= nums[left] && target < nums[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } 
        // Right half is sorted
        else {
            if (target > nums[mid] && target <= nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
    }
    
    return -1;
}

File: 33. Search in Rotated Sorted Array/33. Search in Rotated Sorted Array.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        
        while left <= right {
            let mid = left + (right - left) / 2;
            
            if nums[mid as usize] == target {
                return mid;
            }
            
            // Check if the left half is sorted
            if nums[left as usize] <= nums[mid as usize] {
                if target >= nums[left as usize] && target < nums[mid as usize] {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } 
            // Right half is sorted
            else {
                if target > nums[mid as usize] && target <= nums[right as usize] {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        
        -1
    }
}